gateway/metrics/exporter.go

```diff
@@ -87,40 +132,70 @@ func (e *Exporter) StartServiceWatcher(endpointURL url.URL, metricsOptions Metri
 			ExpectContinueTimeout: 1500 * time.Millisecond,
 		},
 	}
+}
 
-	go func() {
-		for {
-			select {
-			case <-ticker.C:
+func (e *Exporter) getFunctions(endpointURL url.URL, namespace string) ([]types.FunctionStatus, error) {
+	timeout := 3 * time.Second
+	proxyClient := e.getHTTPClient(timeout)
 
-				get, _ := http.NewRequest(http.MethodGet, endpointURL.String()+"system/functions", nil)
-				if e.credentials != nil {
-					get.SetBasicAuth(e.credentials.User, e.credentials.Password)
-				}
+	endpointURL.Path = path.Join(endpointURL.Path, "/system/functions")
+	if len(namespace) > 0 {
+		q := endpointURL.Query()
+		q.Set("namespace", namespace)
+		endpointURL.RawQuery = q.Encode()
+	}
 
-				services := []types.FunctionStatus{}
-				res, err := proxyClient.Do(get)
-				if err != nil {
-					log.Println(err)
-					continue
-				}
-				bytesOut, readErr := ioutil.ReadAll(res.Body)
-				if readErr != nil {
-					log.Println(err)
-					continue
-				}
-				unmarshalErr := json.Unmarshal(bytesOut, &services)
-				if unmarshalErr != nil {
-					log.Println(err)
-					continue
-				}
+	get, _ := http.NewRequest(http.MethodGet, endpointURL.String(), nil)
+	if e.credentials != nil {
+		get.SetBasicAuth(e.credentials.User, e.credentials.Password)
+	}
 
-				e.services = services
+	services := []types.FunctionStatus{}
+	res, err := proxyClient.Do(get)
+	if err != nil {
+		return services, err
+	}
 
-				break
-			case <-quit:
-				return
-			}
-		}
-	}()
+	bytesOut, readErr := ioutil.ReadAll(res.Body)
+	if readErr != nil {
+		return services, readErr
+	}
+
+	unmarshalErr := json.Unmarshal(bytesOut, &services)
+	if unmarshalErr != nil {
+		return services, unmarshalErr
+	}
+	return services, nil
+}
+
+func (e *Exporter) getNamespaces(endpointURL url.URL) ([]string, error) {
+	namespaces := []string{}
+
+	get, _ := http.NewRequest(http.MethodGet, endpointURL.String()+"system/namespaces", nil)
```

Comment:
Don't use string concatenation for path joining, it can be error prone in a way that you will only see at runtime.


This will build the same string safely everytime:
```suggestion
	endpointURL.Path = path.Join(endpointURL.Path, "system", "namespaces")
	get, _ := http.NewRequest(http.MethodGet, endpointURL.String(), nil)
```

Reply:
I will update this as well.